navigation/navigation-runtime/src/wasmJsMain/kotlin/androidx/navigation/BrowserHistory.wasmJs.kt

```diff
@@ -0,0 +1,115 @@
+        var updateState = true
+
+        launch {
+            localWindow.popStateEvents().collect { event ->
+                val state = event.state.toString()
+
+                val restoredRoutes = state.lines()
+                val currentBackStack = navController.currentBackStack.value
+                val currentRoutes = currentBackStack.filter { it.destination !is NavGraph }
+                    .mapNotNull { it.getRouteWithArgs() }
+
+                var commonTail = -1
+                restoredRoutes.forEachIndexed { index, restoredRoute ->
+                    if (index >= currentRoutes.size) {
+                        return@forEachIndexed
+                    }
+                    if (restoredRoute == currentRoutes[index]) {
+                        commonTail = index
+                    }
+                }
+
+                //don't handle next navigation calls
+                updateState = false
+
+                if (commonTail == -1) {
+                    //clear full stack
+                    currentRoutes.firstOrNull()?.let { root ->
+                        navController.popBackStack(root, true)
+                    }
+                } else {
+                    currentRoutes[commonTail].let { lastCommon ->
+                        navController.popBackStack(lastCommon, false)
+                    }
+                }
+
+                //restore stack
+                if (commonTail < restoredRoutes.size - 1) {
+                    val newRoutes = restoredRoutes.subList(commonTail + 1, restoredRoutes.size)
+                    newRoutes.forEach { route -> navController.navigate(route) }
+                }
+            }
+        }
+
+        launch {
+            navController.currentBackStack.collect { stack ->
+                if (stack.isEmpty()) return@collect
+
+                val routes = stack.filter { it.destination !is NavGraph }
+                    .map { it.getRouteWithArgs() ?: return@collect }
+
+                val newUri = "$appAddress/${routes.last()}"
+                val state = routes.joinToString("\n")
+
+
+                if (updateState) {
+                    if (initState) {
+                        localWindow.history.replaceState(state.toJsString(), "", newUri)
+                        initState = false
+                    } else {
+                        localWindow.history.pushState(state.toJsString(), "", newUri)
+                    }
+                }
+                updateState = true
+            }
+        }
+    }
+}
+
+private fun Window.popStateEvents(): Flow<PopStateEvent> = callbackFlow {
+    val localWindow = this@popStateEvents
+    val callback: (Event) -> Unit = { event: Event ->
+        if (!isClosedForSend) {
+            if (event is PopStateEvent) {
+                trySend(event)
+            }
+        }
+    }
+
+    localWindow.addEventListener("popstate", callback)
+    awaitClose {
+        localWindow.removeEventListener("popstate", callback)
```

Comment:
Note: in wasmJs target `removeEventListener` will work correctly only with kotlin 2.0.20 (https://youtrack.jetbrains.com/issue/KT-64565)